src/metabase/lib/schema/expression/arithmetic.cljc

```diff
@@ -33,28 +34,94 @@
 (mr/def ::args.numbers
   [:repeat {:min 2} [:schema [:ref ::expression/number]]])
 
-(defn- plus-minus-schema [tag]
-  [:or
-   [:and
-    [:cat
-     [:= tag]
-     [:schema [:ref ::common/options]]
-     [:schema [:ref ::expression/temporal]]
-     [:repeat {:min 1} [:schema [:ref :mbql.clause/interval]]]]
-    [:fn
-     {:error/message "Temporal arithmetic expression with valid interval units for the expression type"}
-     (fn [[_tag _opts expr & intervals]]
-       (let [expr-type (expression/type-of expr)]
-         (every? #(valid-interval-for-type? % expr-type) intervals)))]]
+(defn- validate-plus-minus-temporal-arithmetic-expression
+  "Validate a `:+` or `:-` expression with temporal args. Return a string describing any errors found, or `nil` if it
+  looks ok."
+  [[_tag _opts & exprs]]
+  (let [{non-intervals false, intervals true} (group-by #(isa? (expression/type-of %) :type/Interval) exprs)]
+    (if-not (= (count non-intervals) 1)
+      "Clause must have exactly one non-interval expression"
+      (let [expr-type (expression/type-of (first non-intervals))]
+        (some (fn [[_tag _opts _n unit :as interval]]
+                (when-not (valid-interval-for-type? interval expr-type)
+                  (str "Cannot add a " unit " interval to a " expr-type " expression")))
+              intervals)))))
+
+(defn- plus-minus-temporal-schema
+  "Create a schema for `:+` or `:-` with temporal args: <temporal> ± <interval(s)> in any order"
+  [tag]
+  [:and
+   {:error/message (str tag " clause with a temporal expression and one or more :interval clauses")}
    [:cat
     [:= tag]
     [:schema [:ref ::common/options]]
-    [:repeat {:min 2} [:schema [:ref ::expression/number]]]]])
+    [:repeat
+     {:min 2}
+     [:or
+      [:schema [:ref ::expression/temporal]]
+      [:schema [:ref :mbql.clause/interval]]]]]
```

Comment:
Could we generate something like `(concat intervals [temporal] intervals)` and say that it has to have at least 2 elements?

Reply:
good idea, I've changed it